### PR TITLE
Added EMAC CMORizer

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR adds a CMORizer for native EMAC model output.

Closes https://github.com/ESMValGroup/ESMValCore/issues/312
Closes https://github.com/ESMValGroup/ESMValCore/issues/309
Closes https://github.com/ESMValGroup/ESMValCore/issues/139
Closes https://github.com/ESMValGroup/ESMValCore/issues/66

Link to documentation: 

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
I encountered a rather awkward problem while implementing this:

EMAC data has lots of variables with `units = "0 - 1"`. While these files can be read by iris (which simply assigns an `invalid_units` attribute), each time a file with this is loaded the message

```bash
ut_scale(): NULL factor argument
ut_are_convertible(): NULL unit argument
```

is printed to the screen. These messages make the log [basically unreadable](https://github.com/ESMValGroup/ESMValCore/files/8652579/out.txt)  if multiple years and/or multiple variables are read. Unfortunately, these messages are not issued via pythons `warnings` module, so they cannot be simply ignored with this.

I really want to get rid of these messages, but I'm not quite sure how. I thought of 3 possible options:

1. Wait until `cf_units` fixes this (see https://github.com/SciTools/cf-units/issues/240).
2. Fix this by redirecting `stderr` in our `load` preprocessor. Unfortunately, [`contextlib.redirect_stderr`](https://docs.python.org/3/library/contextlib.html#contextlib.redirect_stderr) does not work here. I don't fully understand why, but [this answer](https://stackoverflow.com/a/22434262/6864675) here suggests that this function does not work "at file descriptors level". This post gives a rather lengthy alternative that indeed successfully suppresses the message.
3. Fix this by using `fix_file` to remove the faulty units with `netCDF4`.

Option 1  would obviously be the optimal choice, but I'm not sure how long this will take. Option 2 is fast, but I'm not sure if adding this workaround to the heart of our loading mechanism is a good idea. Option 3 is definitely cleaner than option 2, but super slow since it involves copying many files.

Any ideas/recommendations/comments on this? @bouweandela @valeriupredoi @zklaus

**Comment by schlunma** [comment]
Thanks @bouweandela, that's awesome!! I just tested this and works perfectly fine. I will put this into `load()` as a temporary workaround!

**Comment by schlunma** [comment]
> don't think so - the root is `Path(__file__).resolve().parent / 'test_data'` and this is used for at least 4 more tests - cesm2, gfdl etc - why are those not hitting the fan?

`__file__` should be `ESMValCore/tests/integration/cmor/_fixes/conftest.py`, so `Path(__file__).parent` should be `ESMValCore/tests/integration/cmor/_fixes`, right? I think that's correct. And the test is also running fie locally with the correct path...

**Comment by valeriupredoi** [comment]
let's figure it out :grin: I am failing one cesm2 test and forcing to print out where and what's in the test data path

**Comment by bouweandela** [comment]
This message from the checkout step warns about the problem:

> Either git or ssh (required by git to clone through SSH) is not installed in the image. Falling back to CircleCI's native git client but the behavior may be different from official git. If this is an issue, please use an image that has official git and ssh installed.

somehow the emac.nc data file does not get checked out. I will make a pull request to install git and ssh before checkout.

**Comment by valeriupredoi** [comment]
scratch that! am a dummer - it's sequential

**Comment by valeriupredoi** [comment]
aye, @bouweandela is right - no sign of the emac.nc file in the (correct) test data root:

- TEST DATA PATH CESM2 /root/project/tests/integration/cmor/_fixes/test_data
- dir /root/project/tests/integration/cmor/_fixes/test_data contains 'cesm2_cl.nc', 'cesm2_waccm_cl.nc', 'cnrm_cm6_1_cl.nc', 'common_cl_a.nc', 'common_cl_ap.nc', 'common_cl_hybrid_height.nc', 'create_test_data.py', 'gfdl_cm4_cl.nc', 'icon_2d.nc', 'icon_3d.nc', 'icon_grid.nc'
_____________________________ test_sample_data_tas __________

**Comment by bouweandela** [comment]
The problem can indeed be fixed by installing git and ssh before checkout on CircleCI. See #1601 for more info.

**Review by axel-lauer — CHANGES_REQUESTED** [review]
I looked through the plots produced by the EMAC example recipe `recipe_emac_new.yml` and most things look good - as far as I can tell. There are, however, a few items that might need to be double-checked:

* map of precip (given in kg/m2/s): the color scale ranges from 0-10, i.e. the plot is empty and evaluation is not possible.
* map of zg at 500 hPa: the values over Antarctica and over Greenland look very strange, I would expect the values to range somewhere between 5000 and 6000 m but not between 0 and 3000 m as in the plot.
* time series of surface pressure: assuming that the values shown are global mean values, the absolute values and the variability look wrong

**Review by axel-lauer — APPROVED** [review]
I successfully ran `recipe_emac_new.yml` and the output looks fine to me. Three quick answers to my previous comments:

* pr: OK, the preprocessor output looks fine, but I am not sure why the diagnostic expects mm/day instead of simply using the units provided in the metadata. But that's not a problem of the EMAC related additions.
* ps: also OK, the preprocessor output looks fine but I do not fully understand why the diagnostic creates such a strange looking time series. But again, this is not related to the EMAC extensions and thus not relevant here.
* zg @ 500 hPa: OK, the original EMAC output already looks weird with the 500 hPa geopotential heitght in Antarctica and Greenland being lower than the surface height! We should ask our EMAC modelers about this, but again, this is not related to the EMAC additions and thus fine.

So in summary, all fine and green light from the science review.

**Review by remi-kazeroni — APPROVED** [review]
This PR looks good to me: code, documentation, tests, recipe run with `recipe_emac_new.yml`. Thanks for your work @schlunma! It's nice to have support for EMAC data as part of the on-the-fly cmorizers.

**Comment by schlunma** [comment]
Awesome, great!! Many thanks for the quick review @axel-lauer @remi-kazeroni :tada:
